File: 프로그래머스/LV2/코딩테스트 연습/SummerWinter Coding(~2018)/방문 길이/방문 길이.cpp
#include <string>
#include <set>
#include <map>
#define point pair<int,int>
using namespace std;

int solution(string dirs)
{
    int answer = 0;
    int cur_x = 0, cur_y = 0, next_x = 0, next_y = 0;
    set<string> ways;
    string way, reverse_way;

    map<char, point> move = { make_pair('U', point(0,1)),make_pair('R', point(1,0)),
    make_pair('D', point(0,-1)), make_pair('L', point(-1,0)) };

    for (char d : dirs)
    {

        next_x = cur_x + move[d].first;
        next_y = cur_y + move[d].second;

        if (next_x < -5 || next_x > 5 || next_y < -5 || next_y > 5)continue;

        way = to_string(cur_x) + to_string(cur_y) + to_string(next_x) + to_string(next_y);
        reverse_way = to_string(next_x) + to_string(next_y) + to_string(cur_x) + to_string(cur_y);
        if (ways.find(way) == ways.end() && ways.find(reverse_way) == ways.end())
        {
            answer++;
            ways.insert(way);
        }
        cur_x = next_x;
        cur_y = next_y;
    }
    return answer;
}

File: 프로그래머스/LV2/코딩테스트 연습/SummerWinter Coding(~2018)/방문 길이/방문 길이.py
def solution(dirs):
    answer = 0
    ways = set()
    cur_x,cur_y = 0,0
    next_x,next_y = cur_x,cur_y

    move = {'U': (0,1), 'D': (0, -1), 'R': (1, 0), 'L': (-1, 0)}
    for d in dirs:
        next_x = cur_x + move[d][0]
        next_y = cur_y + move[d][1]

        if next_x < -5 or next_x > 5 or next_y < -5 or next_y > 5:
            continue

        way = f"{cur_x}{cur_y}{next_x}{next_y}"
        reverse_way = f"{next_x}{next_y}{cur_x}{cur_y}"
        if way not in ways and reverse_way not in ways:
            answer += 1
            ways.add(way)
        cur_x, cur_y = next_x, next_y
    return answer

